client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -256,34 +307,121 @@ public void close() {
     }
 
     @Override
+    @SneakyThrows
     public Type fetchEvent(EventPointer pointer) throws NoSuchEventException {
         Preconditions.checkNotNull(pointer);
-        // Create SegmentInputStream
+        // Create AsyncSegmentEventReader
         @Cleanup
-        SegmentInputStream inputStream = inputStreamFactory.createInputStreamForSegment(pointer.asImpl().getSegment(),
-                                                                                        pointer.asImpl().getEventLength());
-        inputStream.setOffset(pointer.asImpl().getEventStartOffset());
+        AsyncSegmentEventReader reader = readerFactory.createEventReaderForSegment(
+                pointer.asImpl().getSegment(),
+                AsyncSegmentEventReaderFactory.DEFAULT_BUFFER_SIZE);
+        reader.setOffset(pointer.asImpl().getEventStartOffset());
+        reader.setEndOffset(pointer.asImpl().getEventStartOffset() + pointer.asImpl().getEventLength());
+
         // Read event
         try {
-            ByteBuffer buffer = inputStream.read();
+            ByteBuffer buffer = reader.readAsync().join();
```

Comment:
I suggest calling `get` with a timeout. This may block indefinitely.

Reply:
I agree, `EventStreamReader::fetchEvent` should probably take a timeout, but the existing implementation blocks indefinitely and this PR retains that limitation.  Filed #2673.